### DISCUSSION TITLE
ElectronBot-Studio OSX版

### DISCUSSION THREAD

**Discussion by coolflyreg** [peng-zhihui/ElectronBot#86] [Show and tell]
这是ElectronBot-Studio的苹果系统的编译版本。

链接: https://pan.baidu.com/s/1vQUHCkt6RjYSjy1emR5gtQ 提取码: vwa4 

20220422-1
- 更改库引用方式和编译方式

20220421-1
- 感谢[Eilgnaw](https://github.com/Eilgnaw)提供的opencv的arm64版本dylib。
- 改善了播放视频图片后，逐渐变成喷气机的问题
- 修改了模型面部显示功能：在没有连ElectronBot的时候，可以将播放的图像和摄像头图像，显示到3D模型的屏幕上
- 虽然已经上传了Universal的版本，但是目前阶段，还是建议 M1的OSX系统，使用Rosetta方式打开Intel版本的app

20220420-2
- 合并[Eilgnaw](https://github.com/Eilgnaw)的PR，添加支持m1的arm64的StandaloneFileBrowser。
- 上传了Intel Only的版本，M1的OSX系统，可以试试使用Rosetta方式打开

20220420-1
- 添加多个摄像头切换功能，默认在C++底层，打开第一个摄像头
- 添加点击摄像头图像，打开/关闭 显示摄像头到机器人屏幕的功能

20220417-1
- 添加了摄像头隐私权限调用，第一次打开提示需要访问摄像头，同意以后，退出再次打开，就可以打开摄像头了（在系统偏好设置->安全性与隐私->摄像头 里能看到ElectronBot.app）
- 调通了USB通讯，可以连接硬件了，播放视频也没问题。路径不要有中文，不要有空格

20220406-1
- 修改了屏幕自适应，窗口可以自由缩放了

20220319-1
- 使用UnityStandaloneFileBrowser在OSX系统上打开文件
- 编译通过了USB的接口通信，但是没有STM32F405的板子来验证。
- 修改了在OSX系统的文字显示问题

目前只支持Intel CPU的OSX系统。通过[Eilgnaw](https://github.com/Eilgnaw)的热心测试，M1的OSX也能用了。
开发的系统是OSX 12.3.1

如果遇到未知开发者的问题，在 系统偏好设置 -> 安全与隐私 -> 通用 的下方的提示处，点运行就可以了
如果遇到“已损坏”的提示，运行如下命令（注意命令中的路径）：
```
sudo xattr -rd com.apple.quarantine /Applications/ElectronBot.app
```

**Comment by ChangMenQAQ** [comment]
牛的

**Comment by WillionFe** [comment]
你好，我在电脑上安装了软件，发现一个问题：重新开启软件时，窗口会变得好小😂，求
![Screen Shot 2022-04-02 at 18 50 58 (2)](https://user-images.githubusercontent.com/28861984/161379915-92c94eeb-14d6-4ab5-b0e3-e7da624655db.png)
解

**Reply by coolflyreg** [reply]
我还在另一台mac试了没问题，竟然到你这这样😅

**Reply by WillionFe** [reply]
我又打开来看了一下，窗口变得更小了🤦‍♂️，也可能是我电脑这边的问题

**Reply by WillionFe** [reply]
对了，请教一下大佬有没有办法把“电子”接入homeassistant添加成实体呢？我希望能让它可以跟家里的智能家居互动。

**Reply by coolflyreg** [reply]
这个不清楚。。。我还没还原，离考虑智能家居还早着呢。。。

**Reply by coolflyreg** [reply]
等我在更新一版 :-)

**Comment by peng-zhihui** [comment]
Unity输出需要在player setting里面设置固定分辨率1280x720（编辑器模式也是），因为软件没有做分辨率适配

**Reply by coolflyreg** [reply]
我仿佛记得我设置了。不过也有可能，我尝试做分辨率适配的时候，改掉了，忘了改回来。但是我在另一台mac上就没问题。等我晚上再试试看

**Comment by coolflyreg** [comment]
@WillionFe 我上传了新的版本，网盘地址不变。修改了屏幕自适应，可以缩放屏幕大小了

**Reply by WillionFe** [reply]
可以调整大小了，棒～
<img width="1440" alt="Screen Shot 2022-04-07 at 08 24 25" src="https://user-images.githubusercontent.com/28861984/162096424-827d4159-6bb2-4c9e-9b73-77196135a2c9.png">

**Reply by coolflyreg** [reply]
你第一次打开的时候，会有摄像头权限提示。退出app，再打开，应该就能显示摄像头画面了

**Reply by WillionFe** [reply]
没有权限提示哦，直接就打开了，没有调用电脑自己的摄像头，会不会是因为我还没连ElectronBot的原因

**Reply by coolflyreg** [reply]
你下载的文件名是啥？摄像头其实是opencv调用的。估计之后还得写个选择摄像头的逻辑。
你的是M1的还是Intel的？系统版本是osx 11.x？

**Reply by WillionFe** [reply]
文件名：
ElectronBot-Studio-OSX-2022040701.zip

系统信息：（intel + 12.3）
macOS Monterey
Version 12.3
MacBook Pro (13-inch, 2020, Four Thunderbolt 3 ports)
Processor 2 GHz Quad-Core Intel Core i5
Memory 16 GB 3733 MHz LPDDR4X
Graphics Intel Iris Plus Graphics 1536 MB

**Reply by coolflyreg** [reply]
> 文件名： ElectronBot-Studio-OSX-2022040701.zip
> 
> 系统信息：（intel + 12.3） macOS Monterey Version 12.3 MacBook Pro (13-inch, 2020, Four Thunderbolt 3 ports) Processor 2 GHz Quad-Core Intel Core i5 Memory 16 GB 3733 MHz LPDDR4X Graphics Intel Iris Plus Graphics 1536 MB

再试试新版的

**Comment by WillionFe** [comment]
试了最新版本（ElectronBot-Studio-OSX-2022041701.zip），替换了原来的之后，打开依旧没有摄像头权限提醒，风扇转的比较快了
<img width="1440" alt="Screen Shot 2022-04-17 at 18 08 52" src="https://user-images.githubusercontent.com/28861984/163710089-0cefe0c9-3185-4fa9-a502-868d8a828831.png">

**Reply by coolflyreg** [reply]
奇了怪了，为什么只在我自己的机器上可以。

**Reply by WillionFe** [reply]
我换了台机器测试，依旧没有摄像头权限弹窗，我也不知道为什么😂
macOS Catalina
版本10.15.7
MacBook Air (11-inch,Early 2015)
处理器1.6GHz双核Intel Core i5
![Screen Shot 2022-04-17 at 19 12 11 (2)](https://user-images.githubusercontent.com/28861984/163711977-46d21a96-f5d7-4754-acdd-5a2789c1dced.png)

**Reply by coolflyreg** [reply]
难道得用开发者账号，签名发布后，才会在其他机器上弹吗。。。

**Reply by WillionFe** [reply]
😂不知道耶，要这样子的吗？我只知道ios端是需要这样的

**Reply by coolflyreg** [reply]
你点下那个play左边的log，看看有啥输出

**Reply by WillionFe** [reply]
[I]pzh::Unity::Screen width:640 Screen height:480
<img width="1002" alt="Screen Shot 2022-04-17 at 19 26 34" src="https://user-images.githubusercontent.com/28861984/163712340-83ec5691-5307-4483-957c-5a2e51d6f304.png">

**Reply by coolflyreg** [reply]
我再研究研究。。。我的是这样的
<img width="1099" alt="image" src="https://user-images.githubusercontent.com/4011564/163712591-b33e1809-5dfd-46d4-868f-78472c700a2d.png">

**Reply by WillionFe** [reply]
好嘞，等你好消息～

**Reply by coolflyreg** [reply]
又更新了下，这回加了摄像头切换，再试试，切换下摄像头。测试的时候，除了自带的FaceTime摄像头，至少还得插一个。不然没法触发切换功能

**Reply by coolflyreg** [reply]
这回应该可以了。再试试

**Reply by WillionFe** [reply]
我试了一下,两台电脑都是直接打开了,没有弹窗哦

<img width="1440" alt="Screen Shot 2022-04-27 at 09 06 17" src="https://user-images.githubusercontent.com/28861984/165418142-8ff976af-d1cb-47d9-98ac-97bd0d38292e.png">

**Reply by coolflyreg** [reply]
你这个还是旧版，新的界面有些不同。摄像头窗口下新加了切换摄像头的功能。你这个没有。

**Reply by WillionFe** [reply]
我错了😂,还真是,刚下载下来没替换好,这次OK了,谢谢

<img width="1274" alt="Screen Shot 2022-04-27 at 09 40 13" src="https://user-images.githubusercontent.com/28861984/165421277-d75af333-e0ab-4177-8c6b-da037293b51a.png">

**Reply by coolflyreg** [reply]
good，好消息

**Comment by Eilgnaw** [comment]
感谢大佬. 
另外我在 M1 下测试发现打不开选择文件窗口,我修复了一下,但是我没法测试在 intel 下是否能正常工作,已经给你提 pr 了,方便的话可以测试下.

**Reply by coolflyreg** [reply]
谢谢提交pr，我测试了下，Intel的可以打开。不过我只有intel的机器，其他编译的依赖都只有x86_64架构的。我还没搞明白怎么用cmake编译同时带有x86_64和arm64的。

**Reply by Eilgnaw** [reply]
https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary  分别编译然后合并.另外我看到Plugins 下还有很多 x86_64 的文件,都需要处理一下. 或者你可以把 unity 编译平台设置成 intel, 让m1用户通过 rosetta 打开,这样应该比较省事,不过可能会有问题,由于我没有ElectronBot,所以没办法测试全部功能.

**Reply by coolflyreg** [reply]
这个分开编译后合并我是知道，只是在CMakeList.txt中怎么配置，我还没搞明白。除了这个，其实还有opencv，我用的是macport安装的opencv，dylib也是直接拷贝的macport安装后的文件。由于没有m1机器，我也没法测试，m1的rostta方式打开，其实我也不太明白。不过可以照你说的方法试试

**Reply by coolflyreg** [reply]
我上传了个Intel Only的版本，有空可以试试rosetta打开
